website/src/components/views/wasap/useWasapPageData.ts

```diff
@@ -0,0 +1,116 @@
+        case 'untracked': {
+            const variantsToExclude =
+                analysis.excludeSet === 'custom'
+                    ? analysis.excludeVariants
+                    : await getCladeLineages(
+                          wastewaterConfig.wasap.covSpectrum.lapisBaseUrl,
+                          wastewaterConfig.wasap.covSpectrum.cladeField,
+                          wastewaterConfig.wasap.covSpectrum.lineageField,
+                          true,
+                      ).then((r) => Object.values(r));
+            if (variantsToExclude === undefined) {
+                return { type: 'selected', mutations: [] };
+            }
+            const [excludeMutations, allMuts] = await Promise.all([
+                Promise.all(
+                    variantsToExclude.map((v) =>
+                        getMutations(wastewaterConfig.wasap.covSpectrum.lapisBaseUrl, analysis.sequenceType, v, 0.8, 9),
+                    ),
+                ).then((r) => r.flat()),
+                getMutations(wastewaterConfig.wasap.lapisBaseUrl, analysis.sequenceType, undefined, 0.05, 5),
+            ]);
+            return {
+                type: 'selected',
+                mutations: allMuts.filter((m) => !excludeMutations.includes(m)),
+            };
+        }
+    }
+}
+
+function getDisplayMutationsAndCustomColumns(
+    mutationSelection: MutationSelection | undefined,
+): [string[] | undefined, CustomColumn[] | undefined] {
+    if (mutationSelection === undefined) {
+        return [undefined, undefined];
+    }
+
+    switch (mutationSelection.type) {
+        case 'all':
+            return [undefined, undefined];
+        case 'selected':
+            return [mutationSelection.mutations, undefined];
+        case 'jaccard':
+            return [
+                mutationSelection.mutationsWithScore.map(({ mutation }) => mutation),
+                [
+                    {
+                        header: 'Jaccard index',
+                        values: Object.fromEntries(
+                            mutationSelection.mutationsWithScore.map(({ mutation, jaccardIndex }) => [
+                                mutation,
+                                jaccardIndex.toPrecision(2),
+                            ]),
+                        ),
+                    },
+                ],
+            ];
+    }
+}
+
+/**
+ * Hook that fetches and returns display mutations and custom columns for the WASAP page,
+ * depending on the analyis mode and analysis mode settings.
+ */
+export function useWasapPageData(analysis: WasapAnalysisFilter) {
+    const { data, isPending, isError } = useQuery({
+        queryKey: ['wasap', analysis],
+        queryFn: () => fetchMutationSelection(analysis),
+    });
+
+    const [displayMutations, customColumns] = getDisplayMutationsAndCustomColumns(data);
+
+    return { displayMutations, customColumns, isPending, isError };
```

Comment:
Doesn't matter so much due to how the return value is used, but we lose nice return type of `useQuery` here (i.e. something like: if it's an error, then there must not be data - the caller will not see that. My IDE claims that the return type is `{displayMutations: string[] | undefined, customColumns: {values: Record<string, string | number>, header: string}[] | undefined, isPending: boolean, isError: boolean}`).

I'd suggest to reduce this hook to `return useQuery(...)` and move `getDisplayMutationsAndCustomColumns` into `fetchMutationSelection`.

Reply:
I have modified the function to return the `QueryResult` type directly, by pulling the transforming function into the `queryFn`: https://github.com/GenSpectrum/dashboards/pull/924/commits/3d2416bf6650bd732c91f7ddf7617d1b28701683

I wanted to keep it clean at the call site! Didn't want to have to expose the transform function and internal `MutationSelection` type.

(I also renamed some bit, added some docs)